### PR TITLE
Prmdr 163 migrate authorisation handler

### PR DESCRIPTION
**Pull request by joefong-nhs**
- Add an authoriser lambda which verify and decode a JWT authorizationToken, and respond with IAM policy which allow or deny user access to other APIs
- Add unit test for above lambda
- Change `make zip` command to pack with `--platform manylinux2014_x86_64` flag, as python's cryptography package seems to require some platform specific native library

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -46,7 +46,7 @@ env:
 zip:
 	rm -rf ./lambdas/package_$(lambda_name) || true 
 	mkdir ./lambdas/package_$(lambda_name)
-	./lambdas/venv/bin/pip3 install --cache-dir .pip_cache -r lambdas/requirements.txt -t ./lambdas/package_$(lambda_name)
+	./lambdas/venv/bin/pip3 install --platform manylinux2014_x86_64 --only-binary=:all: --implementation cp  -r lambdas/requirements.txt -t ./lambdas/package_$(lambda_name)
```
Comment (SRAlexander) [inline Makefile:49]: We should give this a good test, but it sounds like a reasonable adjustment.

File: lambdas/handlers/authoriser_handler.py [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+        logger.info(f"decoded JWT: {decoded}")
+    except botocore.exceptions.ClientError as e:
+        logger.error(e)
+    except jwt.PyJWTError as e:
+        logger.error(f"error while decoding JWT: {e}")
+    except KeyError as e:
+        logger.error(e)
+
+    principalId = ""
+    tmp = event["methodArn"].split(":")
```
Comment (SRAlexander) [inline lambdas/handlers/authoriser_handler.py:53]: can we rename this to something more meaningful?

File: lambdas/handlers/authoriser_handler.py [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+import botocore
+import jwt
+
+logger = logging.getLogger()
+logger.setLevel(logging.INFO)
+
+
+def lambda_handler(event, context):
+    user = "Unauthorised"
+    ssm_public_key_parameter_name = os.environ["SSM_PARAM_JWT_TOKEN_PUBLIC_KEY"]
```
Comment (SRAlexander) [inline lambdas/handlers/authoriser_handler.py:27]: How are we planning to set the os.env on the aws lambda instance?

**Review by SRAlexander — CHANGES_REQUESTED** [review]
Looks good, we will have to look at the deployment of lambda but we will have to wait until we can merge the infrastructure side.

File: lambdas/handlers/authoriser_handler.py (hunk 2 above)
Comment (joefong-nhs) [reply lambdas/handlers/authoriser_handler.py]: Thanks!  Addressed in this commit: https://github.com/nhsconnect/national-document-repository/pull/42/commits/c0192931217fdc6f88a751322e92b745595cfa75

File: lambdas/handlers/authoriser_handler.py (hunk 3 above)
Comment (joefong-nhs) [reply lambdas/handlers/authoriser_handler.py:27]: In a branch for subtask 165, we added this env var to the lambda by terraform : https://github.com/nhsconnect/national-document-repository-infrastructure/blob/prmdr_165_subtask/infrastructure/lambda-authoriser.tf

**Pull request merged**